Content.Server/Fluids/EntitySystems/MoppingSystem.cs

```diff
@@ -0,0 +1,363 @@
+        SubscribeLocalEvent<AbsorbentComponent, AfterInteractEvent>(OnAfterInteract);
+        SubscribeLocalEvent<TransferCancelledEvent>(OnTransferCancelled);
+        SubscribeLocalEvent<TransferCompleteEvent>(OnTransferComplete);
+    }
+
+    private void OnAfterInteract(EntityUid uid, AbsorbentComponent component, AfterInteractEvent args)
+    {
+        var user = args.User;
+        var used = args.Used;
+        var target = args.Target;
+
+        _solutionSystem.TryGetSolution(used, "absorbed", out var absorbedSolution);
+
+        var toolAvailableVolume = FixedPoint2.New(0);
+        var toolCurrentVolume = FixedPoint2.New(0);
+        var transferAmount = FixedPoint2.New(0);
+
+        if (absorbedSolution is not null)
+        {
+            toolAvailableVolume = absorbedSolution.AvailableVolume;
+            toolCurrentVolume = absorbedSolution.CurrentVolume;
+        }
+
+        if (!args.CanReach)
+        {
+            return;
+        }
+
+        // For adding liquid to an empty floor tile
+        if (target is null // if a tile is clicked
+            && !args.Handled)
+        {
+            ReleaseToFloor(args.ClickLocation, component, absorbedSolution);
+            args.Handled = true;
+            return;
+        }
+        else if (target is not null)
+        {
+            // Handle our do_after logic
+            HandleDoAfter(user, used, target.Value, component, toolCurrentVolume, toolAvailableVolume);
+        }
+
+        args.Handled = true;
+        return;
+    }
+
+
+    private void ReleaseToFloor(EntityCoordinates clickLocation, AbsorbentComponent absorbent, Solution? absorbedSolution)
+    {
+
+        if ((_mapManager.TryGetGrid(clickLocation.GetGridId(EntityManager), out var mapGrid)) // needs valid grid
+            && absorbedSolution != null) // needs a solution to place on the tile
+        {
+            TileRef tile = mapGrid.GetTileRef(clickLocation);
+
+            // Drop some of the absorbed liquid onto the ground
+            var releaseAmount = FixedPoint2.Min(absorbent.ResidueAmount, absorbedSolution.CurrentVolume);
+            var releasedSolution = _solutionSystem.SplitSolution(absorbent.Owner, absorbedSolution, releaseAmount);
+            _spillableSystem.SpillAt(tile, releasedSolution, "PuddleSmear");
+        }
+        return;
+    }
+
+    // Handles logic for our different types of valid target.
+    // Checks for conditions that would prevent a doAfter from starting.
+    private void HandleDoAfter(EntityUid user, EntityUid used, EntityUid target, AbsorbentComponent component, FixedPoint2 currentVolume, FixedPoint2 availableVolume)
+    {
+        // Below variables will be set within this function depending on what kind of target was clicked.
+        // They will be passed to the OnTransferComplete if the doAfter succeeds.
+
+        EntityUid donor = new(0);
+        EntityUid acceptor = new(0);
+        var donorSolutionName = "";
+        var acceptorSolutionName = "";
```

Comment:
just dont initialize these tbh

Reply:
this includes the ones below

Reply:
I went with just declaring them here instead of declaring + initializing to dummy blank values. However depending on how we do this, some of these variables may not belong in this function at all.